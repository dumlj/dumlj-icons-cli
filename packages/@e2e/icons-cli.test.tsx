import { describe, expect } from '@jest/globals'
import { SpawnOptions, spawn } from 'child_process'
import path from 'path'
import renderer from 'react-test-renderer'
import fs from 'fs'
import React from 'react'

/** 执行命令, 兼容 spawn 无法捕获错误 */
export const spawnCommand = (command: string, args?: readonly string[], options?: SpawnOptions) => {
  return new Promise((resolve, reject) => {
    const content = []
    const errors = []
    const cp = spawn(command, args, options)
    const { stdout, stderr } = cp

    stderr.on('data', (chunk) => errors.push(chunk.toString()))
    stdout.on('data', (chunk) => content.push(chunk.toString()))

    cp.on('error', (err) => {
      console.log(err)
      reject(err)
    })

    cp.on('close', () => {
      if (errors.length > 0) {
        reject(new Error(errors.join('')))
        return
      }

      resolve(content.join(''))
    })
  })
}

describe('test icon-cli', () => {
  const project = path.join(__dirname, './icons-lib')

  const genIcon = () => {
    return spawnCommand('dumlj-icons', ['-c', './icons.config.js'], {
      cwd: project,
    })
  }

  const buildIcon = () => {
    return spawnCommand('babel', ['./src', '--out-dir', 'dist', '--extensions', '.js'], {
      cwd: project,
    })
  }

  beforeAll(async () => {
    await genIcon()

    await buildIcon()
  }, 120 * 1000)

  afterAll(() => {
    spawnCommand('rm', ['-rf', 'src'], {
      cwd: project,
    })
  })

  it('icons renders correctly', async () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    const { default: Bell } = await import('./icons-lib/dist/icons/Bell')
    const tree = renderer.create(<Bell />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('generate icons folders correctly', () => {
    const checkFolderNames = () => {
      const iconsFolder = fs.existsSync(path.join(project, 'src/icons'))
      const styleFolder = fs.existsSync(path.join(project, 'src/style'))
      const svgFolder = fs.existsSync(path.join(project, 'src/svg'))
      const utilsFolder = fs.existsSync(path.join(project, 'src/utils'))

      return iconsFolder && styleFolder && svgFolder && utilsFolder
    }

    expect(checkFolderNames()).toBeTruthy()
  })
})
