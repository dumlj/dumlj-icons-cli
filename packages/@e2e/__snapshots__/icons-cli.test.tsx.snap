// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test icon-cli icons renders correctly 1`] = `
<svg
  data-icon="Bell"
  fill="currentColor"
  height="1em"
  style={{}}
  viewBox="0 0 24 24"
  width="1em"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    clipRule="evenodd"
    d="M12 1.887a.9.9 0 0 1 .9.9v1.566A6.509 6.509 0 0 1 18.51 10.8v4.006a1.504 1.504 0 0 0 1.504 1.504.9.9 0 0 1 0 1.8H3.988a.9.9 0 0 1 0-1.8 1.504 1.504 0 0 0 1.504-1.504V10.8A6.509 6.509 0 0 1 11.1 4.353V2.787a.9.9 0 0 1 .9-.9ZM6.93 16.31h10.14a3.305 3.305 0 0 1-.361-1.504V10.8a4.709 4.709 0 0 0-9.417 0v4.006c0 .528-.126 1.043-.362 1.504Zm3.233 3.326a.9.9 0 0 1 1.23.327.703.703 0 0 0 1.215 0 .9.9 0 0 1 1.557.903 2.502 2.502 0 0 1-4.33 0 .9.9 0 0 1 .328-1.23Z"
    fillRule="evenodd"
  />
</svg>
`;

exports[`test icon-cli renders correctly 1`] = `
<svg
  data-icon="Bell"
  fill="currentColor"
  height="1em"
  style={{}}
  viewBox="0 0 24 24"
  width="1em"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    clipRule="evenodd"
    d="M12 1.887a.9.9 0 0 1 .9.9v1.566A6.509 6.509 0 0 1 18.51 10.8v4.006a1.504 1.504 0 0 0 1.504 1.504.9.9 0 0 1 0 1.8H3.988a.9.9 0 0 1 0-1.8 1.504 1.504 0 0 0 1.504-1.504V10.8A6.509 6.509 0 0 1 11.1 4.353V2.787a.9.9 0 0 1 .9-.9ZM6.93 16.31h10.14a3.305 3.305 0 0 1-.361-1.504V10.8a4.709 4.709 0 0 0-9.417 0v4.006c0 .528-.126 1.043-.362 1.504Zm3.233 3.326a.9.9 0 0 1 1.23.327.703.703 0 0 0 1.215 0 .9.9 0 0 1 1.557.903 2.502 2.502 0 0 1-4.33 0 .9.9 0 0 1 .328-1.23Z"
    fillRule="evenodd"
  />
</svg>
`;
